zonec.c

```diff
@@ -1215,6 +1599,116 @@ zadd_rdata_txt_clean_wireformat()
 	}
 }
 
+static int
+svcparam_key_cmp(const void *a, const void *b)
+{
+	return ((int)read_uint16(rdata_atom_data(*(rdata_atom_type *)a)))
+	     - ((int)read_uint16(rdata_atom_data(*(rdata_atom_type *)b)));
+}
+
+void
+zadd_rdata_svcb_check_wireformat(uint16_t rr_type)
+{
+	size_t i, j;
+	uint8_t paramkeys[SVCPARAMKEY_COUNT - 1];
+	memset(paramkeys, 0, sizeof(paramkeys));
+	uint16_t key;
+
+	/* 
+	 * In draft-ietf-dnsop-svcb-https-04 Section 7:
+	 * In wire format, the keys are represented by their numeric values in
+	 * network byte order, concatenated in ascending order.
+	 */
+	qsort( (void *)&parser->current_rr.rdatas[2]
+	     , parser->current_rr.rdata_count - 2
+	     , sizeof(rdata_atom_type *)
+	     , svcparam_key_cmp
+	     );
+
+	for (i = 2; i < parser->current_rr.rdata_count; i++) {
+		// @TODO remove after debugging is done
+		// size_t    size = rdata_atom_size(parser->current_rr.rdatas[i]);
+
+		uint8_t  *data = rdata_atom_data(parser->current_rr.rdatas[i]);
+		key  = read_uint16(data);
+
+		/* keep track of keys that are present*/
+		if (key < SVCPARAMKEY_COUNT) {
+			paramkeys[key] = 1;
+		}
+
+		// @TODO check doubles
+
+		// fprintf(stderr, "SvcParam[%zu]: len: %zu, type:", i, size);
+		// if (key < SVCPARAMKEY_COUNT)
+		// 	fprintf(stderr, "%s\n", svcparamkey_strs[key]);
+		// else
+		// 	fprintf(stderr, "key%hu\n", key);
+	}
+
+	if (paramkeys[SVCB_KEY_ALPN] && paramkeys[SVCB_KEY_NO_DEFAULT_ALPN])
+		zc_error_prev_line("alpn and no-default-alpn are not allowed"
+			                   " in the same record");
+
+	/*
+	 * In draft-ietf-dnsop-svcb-https-04 Section 8, Using SVCB with HTTPS and HTTP:
+	 * The "automatically mandatory" keys (Section 7) are "port"
+	 * and "no-default-alpn". 
+	 * (...) 
+	 * "automatically mandatory", i.e. mandatory if they arepresent in an RR.
+	 */
+	if (rr_type == TYPE_HTTPS) {
+		if (!(paramkeys[SVCB_KEY_PORT]) || !(paramkeys[SVCB_KEY_NO_DEFAULT_ALPN])) {
+			zc_error_prev_line("port and no-default-alpn must be included"
+			                   " in the HTTPS record");
+		}
+	}
```

Comment:
```suggestion
	if (rr_type == TYPE_HTTPS
	&& (!paramkeys[SVCB_KEY_PORT] || !paramkeys[SVCB_KEY_NO_DEFAULT_ALPN])) {
		zc_error_prev_line("port and no-default-alpn must be included"
			                   " in the HTTPS record");
	}
```
I think leaving out unnecessary nesting and brackets increase readability, but feel free to disagree.